### DISCUSSION TITLE
Speed Comparison

### DISCUSSION THREAD

**Discussion by Baron-von-Riedesel** [dosemu2/dosemu2#1875] [General]
Since DosEmu2 has become so stable recently, I did a little "benchmark" to see how it compares to other environments.
The results are quite interesting:

```
 Speed test assembling debug.asm to debugx.com with JWasm,
 results in ms. Option -Fl creates debugx.lst ( 1.3 MB ).

 environ                     with -Fl      w/o -Fl  comment
 -----------------------------------------------------------------------------
 MS-DOS 7.1, SmartDrv          265            30
 Linux                         105            43    64-bit OS, 64-bit JWasm
 DosEmu2 (KVM, Drive C:)      5950            43    DosEmu2 w/o debug info
 Qemu (KVM),DOS7.1            2150            70 
 Qemu (KVM),DOS7.1,SmartDrv    550            55 
 86Box                        6350          2000


For DosEmu2 and 86Box, loading SmartDrive gives no significant change.
```

- DoeEmu beats Qemu if there's not much output to a file.
- with lots of file output, DosEmu slows down so significantly that it almost has to struggle to beat 86Box.
- native DOS is still very competitive. :)

**Comment by stsp** [comment]
Are you running dosemu2 with kvm?
"w/o debug info" - what does it mean?
debug info is currently added even for
release builds.
Please consider creating an easy to
run test-case.

**Comment by andrewbird** [comment]
> DosEmu2 (KVM, Drive C:)  

Please could you repeat that test on a different drive that is a fat image file?

You also need to mention if you are using FDPP or FreeDOS etc.

**Comment by Baron-von-Riedesel** [comment]
> Are you running dosemu2 with kvm?

guess what "DosEmu2 (KVM, Drive C:)" could probably mean...

> debug info is currently added even for release builds.

I noticed that - so I changed file Makefile.conv manually; but that was before I realized that the slowing factor is file i/o and debug info isn't really relevant here.

> Please consider creating an easy to run test-case. 

This is not a bug report. But the test case is indeed pretty "easy to run", so it's attached ( make.bat runs with -Fl, make2.bat runs without -Fl )

[DEBUG.zip](https://github.com/dosemu2/dosemu2/files/10368313/DEBUG.zip)

> You also need to mention if you are using FDPP or FreeDOS etc.

Yes, FDPP used.

**Comment by stuaxo** [comment]
Maybe a silly question:. Does freedos have some equivalent of smartdrv ?

I wonder if it would make the results compatible to DOS with smartdrv ?

Still - It's not something I'd recommend in real usage.

**Comment by stsp** [comment]
I applied the patch that improves
the situation by quarter. Still not good.
You need to disable KVM to get a
better speed.
The problem seems to be hpet.
With perf I am getting that trace:
```
   - 57,61% kvm_run                                                            ▒
      - 57,46% __GI___ioctl (inlined)                                          ▒
         - 43,75% entry_SYSCALL_64_after_hwframe                               ▒
            - 43,51% do_syscall_64                                             ▒
               - 41,58% __x64_sys_ioctl                                        ▒
                  - 40,84% kvm_vcpu_ioctl                                      ▒
                     - 40,24% kvm_arch_vcpu_ioctl_run                          ▒
                        - 33,84% vcpu_run                                      ▒
                           - 33,57% vcpu_enter_guest                           ▒
                              - 21,03% kvm_guest_time_update                   ▒
                                 - 20,37% ktime_get_raw                        ▒
                                      read_hpet
```

Maybe @seanjc can tell us why
kvm+hpet is so slow - slower than
our home-brew cpu simulator?

**Reply by stsp** [reply]
Mis-spelled, sorry: @sean-jc

**Comment by stsp** [comment]
I added a few optimizations.
So now you need to disable KVM:
```
$_cpu_vm = "emulated"
$_cpu_vm_dpmi = "native"
```
and let us know the results.
After KVM, the slow-down factor
was EMM. Which is known to be
very slow on old PCs, where it was
present on an external boards. We
emulate that property very well. :)
But now it is avoided for small I/O
blocks. I've got a performance almost
twice as good as with dosbox, and
that is w/o kvm!
Of course it would be cool to find
out how to unbind kvm from slow
hpet...

**Reply by Baron-von-Riedesel** [reply]
> and let us know the results.

Yes, a really nice improvement:
```
 Speed test assembling debug.asm to debugx.com with JWasm,
 results in ms. Option -Fl creates debugx.lst ( 1.3 MB ).

 environ                     with -Fl      w/o -Fl  comment
 -----------------------------------------------------------------------------
 MS-DOS 7.1, SmartDrv          265            30
 Linux                         105            43    64-bit OS, 64-bit JWasm
 DosEmu2 (KVM, Drive C:)      5950            43
 DosEmu2 (emulated/native)    3850            43
 Qemu (KVM),DOS7.1            2150            70 
 Qemu (KVM),DOS7.1,SmartDrv    550            55 
 86Box                        6350          2000


 DosEmu2 source 10.1.2023:

 environ                     with -Fl      w/o -Fl  comment
 -----------------------------------------------------------------------------
 DosEmu2 (KVM)                3620            43
 DosEmu2 (emulated/native)    2180            43
```

**Comment by stsp** [comment]
Added "tsc=reliable" to kernel cmdline
and rebooted. kvm now twice faster than
before. But still can't compete with cpu-emu
in your test. So disabling KVM is still the
need.

**Comment by stsp** [comment]
> For DosEmu2 and 86Box, loading SmartDrive gives no significant change.

For dosemu2 you need to write
your own smartdrive that hooks
int21 in prot mode, rather than
int13 in real mode. Then it will
work with your test-case as
expected.

**Comment by stuaxo** [comment]
@Baron-von-Riedesel it might be interesting to add FreeDOS, native to the list + MSDOS without smartdrv.   I found Mateusz Vistes old freedos bug from 2015 that implies there is a tiny bit of caching in MSDOS that isn't in FreeDOS (though in that case it was an older FreeDOS and and MSDOS 3.3)

https://sourceforge.net/p/freedos/mailman/freedos-user/thread/CANv4V5jGPDxXqt7NK5R8rGUA8DD6tvG%2Bs%3DgdE0xG%2BsEQLDpf%2BA%40mail.gmail.com/

**Reply by stsp** [reply]
That post is likely very specific to
floppies. freedos also has the concept
of "buffers", just as ms-dos. Mateusz
probably only says that they forgot to
apply a few of those buffers to floppy.

Ultimately, the test-case here does
zillions of very small writes. Under
qemu this is cured by smartdrive,
because there you get the kvm exit
per every file write.
But under dosemu2 you have a
kvm exit to handle prot mode call,
then switch to realmode, then exit
for a file write - at least twice more
vmexits.

**Reply by stuaxo** [reply]
Ah, I see - that makes the bit about needing a protected mode disk cache make sense.

**Comment by Baron-von-Riedesel** [comment]
> it might be interesting to add FreeDOS, native to the list 

FreeDOS times are similiar to MS-DOS, generally it's a bit slower, but nothing interesting here.

> MSDOS without smartdrv

That does increase times a lot, to the 6000 ms "region".

> Ultimately, the test-case here does zillions of very small writes. 

Not exactly "zillions" :), but quite a few, that's right - after all, it uses C's "buffered" file I/O, and I guess that has been "invented" just to cure this "stupid" behavior. If the file buffer would be increased to 4 kB ( don't know the current default value of Open Watcom ), things might improve drastically.

**Reply by stsp** [reply]
But not for dosemu2, as in that
case you would need buffering
before the protected mode int21
handler. Try to write such
smartdrive. :)

**Reply by Baron-von-Riedesel** [reply]
But the "buffered" I/O" code - as part of the CRT -   DOES run in protected-mode, so it should smoothen the many small writes.
However, I checked it, Open Watcom's default buffer size already IS 4096 bytes, so the effect obviously isn't sufficient,
Btw., replacing jwasm by masm gives much better results for DosEmu2 - apparently masm is optimized in this regard ( not really surprising, since masm was designed and written in the 1980s ).

**Comment by stsp** [comment]
> But the "buffered" I/O" code - as part of the CRT - DOES run in protected-mode, so it should smoothen

If you call the buffered IO funcs
explicitly. But in your test-case
there are direct int21 calls in prot
mode.

**Reply by Baron-von-Riedesel** [reply]
> But in your test-case there are direct int21 calls in prot mode.

That's just because the HX Win32 emulation code is linked staticfally into the binary to make it stand-alone. That is, the jwasm C code is linked with the OW CRT for Windows NT, and since that code links to the Win32 API, the HX Win32 emulation layer is added by the linker. The jwasm code just uses the C runtime functions ( else it surely wouldn't be so easy to create linux/maxOS versions ).
